File: scripts/gate5-load.py
import concurrent.futures
import json
import os
import statistics
import time
import urllib.error
import urllib.request
from pathlib import Path

BASE_URL = os.environ['GATE5_CANARY_URL'].rstrip('/')
TOKEN = os.environ['GATE5_CANARY_TOKEN']
OUT = Path(os.environ.get('GATE5_OUTPUT_DIR', 'gate5-artifacts'))
OUT.mkdir(parents=True, exist_ok=True)

PERSONAS = [
    'emma_usa','oliver_uk','liam_australia','minji_korea','pavel_belarus','lukas_germany','aina_malaysia','dimas_indonesia',
    'bence_hungary','yuting_taiwan','zofia_poland','matas_lithuania','ananya_india','xinyi_china','linh_vietnam','rahul_bangladesh',
    'nadeesha_srilanka','suman_nepal','amara_nigeria','andrei_romania',
]
LEVELS = [10, 20, 30, 35]


def post_json(path, payload, headers=None, timeout=50):
    data = json.dumps(payload).encode('utf-8')
    req_headers = {'Content-Type': 'application/json'}
    if headers:
        req_headers.update(headers)
    req = urllib.request.Request(BASE_URL + path, data=data, headers=req_headers, method='POST')
    started = time.perf_counter()
    try:
        with urllib.request.urlopen(req, timeout=timeout) as response:
            body = response.read()
            elapsed = (time.perf_counter() - started) * 1000
            return response.status, dict(response.headers.items()), body, elapsed
    except urllib.error.HTTPError as exc:
        body = exc.read()
        elapsed = (time.perf_counter() - started) * 1000
        return exc.code, dict(exc.headers.items()), body, elapsed


def run_flow(index):
    persona = PERSONAS[index % len(PERSONAS)]
    total_started = time.perf_counter()
    chat_status, chat_headers, chat_body, chat_ms = post_json('/api/chat', {
        'message': 'Hello! Nice to meet you.',
        'history': [],
        'topic': 'intro',
        'aiStudentId': persona,
    })
    result = {
        'persona': persona,
        'chat_status': chat_status,
        'chat_ms': round(chat_ms, 1),
        'tts_status': None,
        'tts_ms': None,
        'combined_ms': None,
        'provider': None,
        'voice': None,
        'ok': False,
        'error': '',
    }
    if chat_status != 200:
        result['error'] = f'chat_http_{chat_status}:{chat_body[:200].decode("utf-8", "replace")}'
        result['combined_ms'] = round((time.perf_counter() - total_started) * 1000, 1)
        return result
    try:
        parsed = json.loads(chat_body)
        if parsed.get('success') is not True or parsed.get('route') != 'anthropic-resilient':
            result['error'] = 'chat_route_or_success_invalid'
            result['combined_ms'] = round((time.perf_counter() - total_started) * 1000, 1)
            return result
        reply = str(parsed.get('data', {}).get('reply') or '').strip()
        if not reply:
            result['error'] = 'chat_reply_empty'
            result['combined_ms'] = round((time.perf_counter() - total_started) * 1000, 1)
            return result
    except Exception as exc:
        result['error'] = f'chat_json_invalid:{exc}'
        result['combined_ms'] = round((time.perf_counter() - total_started) * 1000, 1)
        return result

    tts_status, tts_headers, tts_body, tts_ms = post_json('/api/tts/azure-canary', {
        'text': reply[:280],
        'aiStudentId': persona,
        'speakingRate': 0.95,
    }, headers={'X-Azure-Canary-Token': TOKEN}, timeout=25)
    result['tts_status'] = tts_status
    result['tts_ms'] = round(tts_ms, 1)
    result['provider'] = tts_headers.get('X-TTS-Provider') or tts_headers.get('x-tts-provider')
    result['voice'] = tts_headers.get('X-TTS-Voice') or tts_headers.get('x-tts-voice')
    result['combined_ms'] = round((time.perf_counter() - total_started) * 1000, 1)
    if tts_status != 200:
        result['error'] = f'tts_http_{tts_status}:{tts_body[:200].decode("utf-8", "replace")}'
        return result
    if result['provider'] != 'azure-speech':
        result['error'] = f'provider_mismatch:{result["provider"]}'
        return result
    if len(tts_body) < 500:
        result['error'] = f'tts_audio_too_small:{len(tts_body)}'
        return result
    result['ok'] = True
    return result


def percentile(values, q):
    if not values:
        return None
    ordered = sorted(values)
    idx = max(0, min(len(ordered) - 1, int(round((len(ordered) - 1) * q))))
    return round(ordered[idx], 1)


all_results = []
level_summaries = []
flow_counter = 0
for level in LEVELS:
    started = time.perf_counter()
    with concurrent.futures.ThreadPoolExecutor(max_workers=level) as pool:
        futures = [pool.submit(run_flow, flow_counter + i) for i in range(level)]
        results = [future.result() for future in futures]
    flow_counter += level
    wall_ms = round((time.perf_counter() - started) * 1000, 1)
    all_results.extend(results)
    combined = [row['combined_ms'] for row in results if row['combined_ms'] is not None]
    chat = [row['chat_ms'] for row in results if row['chat_ms'] is not None]
    tts = [row['tts_ms'] for row in results if row['tts_ms'] is not None]
    errors = [row for row in results if not row['ok']]
    level_summaries.append({
        'concurrency': level,
        'attempted': len(results),
        'succeeded': sum(1 for row in results if row['ok']),
        'failed': len(errors),
        'http_429': sum(1 for row in results if row['chat_status'] == 429 or row['tts_status'] == 429),
        'chat_p50_ms': percentile(chat, 0.50),
        'chat_p95_ms': percentile(chat, 0.95),
        'tts_p50_ms': percentile(tts, 0.50),
        'tts_p95_ms': percentile(tts, 0.95),
        'combined_p50_ms': percentile(combined, 0.50),
        'combined_p95_ms': percentile(combined, 0.95),
        'wall_ms': wall_ms,
    })
    print(json.dumps(level_summaries[-1], ensure_ascii=False))

summary = {
    'gate': 5,
    'service': BASE_URL,
    'levels': level_summaries,
    'attempted_flows': len(all_results),
    'succeeded_flows': sum(1 for row in all_results if row['ok']),
    'failed_flows': sum(1 for row in all_results if not row['ok']),
    'http_429_total': sum(1 for row in all_results if row['chat_status'] == 429 or row['tts_status'] == 429),
    'providers': sorted({row['provider'] for row in all_results if row['provider']}),
    'voices_observed': len({row['voice'] for row in all_results if row['voice']}),
    'firestore_session_write_calls': 0,
}
(OUT / 'gate5-summary.json').write_text(json.dumps(summary, indent=2, ensure_ascii=False))
(OUT / 'gate5-flows.json').write_text(json.dumps(all_results, indent=2, ensure_ascii=False))

expected = sum(LEVELS)
if summary['attempted_flows'] != expected:
    raise SystemExit(f'Gate 5 FAIL: attempted {summary["attempted_flows"]}, expected {expected}')
if summary['failed_flows'] != 0:
    first = next(row for row in all_results if not row['ok'])
    raise SystemExit(f'Gate 5 FAIL: {summary["failed_flows"]} failed flows; first={first}')
if summary['http_429_total'] != 0:
    raise SystemExit(f'Gate 5 FAIL: HTTP 429 count={summary["http_429_total"]}')
if summary['providers'] != ['azure-speech']:
    raise SystemExit(f'Gate 5 FAIL: provider set={summary["providers"]}')
if any((level['combined_p95_ms'] or 0) > 30000 for level in level_summaries):
    raise SystemExit('Gate 5 FAIL: combined p95 exceeded 30 seconds')

print('Gate 5: PASS')
print(json.dumps(summary, indent=2, ensure_ascii=False))

File: scripts/gate5-load.ts
import fs from 'node:fs/promises';
import { synthesizeAzureTts } from '../src/server/azureTts';
import { TARGET_20_AI_STUDENT_IDS } from '../src/data/curriculum';

const chatBaseUrl = (process.env.GATE5_CHAT_URL || 'https://shizuoka-english-ai-1075707511474.asia-northeast1.run.app').replace(/\/$/, '');
const outputDir = process.env.GATE5_OUTPUT_DIR || 'gate5-artifacts';
const levels = [10, 20, 30, 35] as const;
const timeoutMs = 50_000;

type FlowResult = {
  personaId: string;
  chatStatus: number;
  ttsStatus: 'ok' | 'error' | 'skipped';
  chatMs: number;
  ttsMs: number | null;
  combinedMs: number;
  provider: string | null;
  voice: string | null;
  ok: boolean;
  error: string;
};

function percentile(values: number[], q: number): number | null {
  if (!values.length) return null;
  const ordered = [...values].sort((a, b) => a - b);
  const index = Math.max(0, Math.min(ordered.length - 1, Math.round((ordered.length - 1) * q)));
  return Math.round(ordered[index] * 10) / 10;
}

async function runFlow(index: number): Promise<FlowResult> {
  const personaId = TARGET_20_AI_STUDENT_IDS[index % TARGET_20_AI_STUDENT_IDS.length];
  const totalStart = performance.now();
  const chatStart = performance.now();
  let chatResponse: Response;
  try {
    chatResponse = await fetch(`${chatBaseUrl}/api/chat`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        message: 'Hello! Nice to meet you.',
        history: [],
        topic: 'intro',
        aiStudentId: personaId,
      }),
      signal: AbortSignal.timeout(timeoutMs),
    });
  } catch (error) {
    const now = performance.now();
    return { personaId, chatStatus: 0, ttsStatus: 'skipped', chatMs: now - chatStart, ttsMs: null, combinedMs: now - totalStart, provider: null, voice: null, ok: false, error: `chat_fetch:${error instanceof Error ? error.message : String(error)}` };
  }
  const chatMs = performance.now() - chatStart;
  const chatBody = await chatResponse.text();
  if (!chatResponse.ok) {
    return { personaId, chatStatus: chatResponse.status, ttsStatus: 'skipped', chatMs, ttsMs: null, combinedMs: performance.now() - totalStart, provider: null, voice: null, ok: false, error: `chat_http_${chatResponse.status}:${chatBody.slice(0, 160)}` };
  }

  let reply = '';
  try {
    const parsed = JSON.parse(chatBody) as {
      success?: boolean;
      isFallback?: boolean;
      data?: { reply?: string };
      _diagnostics?: { route?: string; model?: string };
    };
    const route = parsed._diagnostics?.route || '';
    if (parsed.success !== true || parsed.isFallback !== false || route !== 'anthropic-resilient') {
      throw new Error(`route=${route},success=${String(parsed.success)},fallback=${String(parsed.isFallback)}`);
    }
    reply = String(parsed.data?.reply || '').trim();
    if (!reply) throw new Error('empty reply');
  } catch (error) {
    return { personaId, chatStatus: chatResponse.status, ttsStatus: 'skipped', chatMs, ttsMs: null, combinedMs: performance.now() - totalStart, provider: null, voice: null, ok: false, error: `chat_invalid:${error instanceof Error ? error.message : String(error)}` };
  }

  const ttsStart = performance.now();
  try {
    const tts = await synthesizeAzureTts(reply.slice(0, 280), personaId, 0.95, 20_000);
    const ttsMs = performance.now() - ttsStart;
    return {
      personaId,
      chatStatus: chatResponse.status,
      ttsStatus: 'ok',
      chatMs,
      ttsMs,
      combinedMs: performance.now() - totalStart,
      provider: tts.provider,
      voice: tts.voiceName,
      ok: tts.audio.length > 500 && tts.provider === 'azure-speech',
      error: tts.audio.length > 500 && tts.provider === 'azure-speech' ? '' : `tts_invalid:${tts.audio.length}:${tts.provider}`,
    };
  } catch (error) {
    const ttsMs = performance.now() - ttsStart;
    return { personaId, chatStatus: chatResponse.status, ttsStatus: 'error', chatMs, ttsMs, combinedMs: performance.now() - totalStart, provider: null, voice: null, ok: false, error: `tts_error:${error instanceof Error ? error.message : String(error)}` };
  }
}

await fs.rm(outputDir, { recursive: true, force: true });
await fs.mkdir(outputDir, { recursive: true });

const all: FlowResult[] = [];
const levelSummaries: Record<string, unknown>[] = [];
let offset = 0;
for (const concurrency of levels) {
  const started = performance.now();
  const results = await Promise.all(Array.from({ length: concurrency }, (_, i) => runFlow(offset + i)));
  offset += concurrency;
  all.push(...results);
  const chat = results.map((r) => r.chatMs);
  const tts = results.map((r) => r.ttsMs).filter((v): v is number => typeof v === 'number');
  const combined = results.map((r) => r.combinedMs);
  const summary = {
    concurrency,
    attempted: results.length,
    succeeded: results.filter((r) => r.ok).length,
    failed: results.filter((r) => !r.ok).length,
    http429: results.filter((r) => r.chatStatus === 429 || r.error.includes('HTTP_429') || r.error.includes('HTTP 429')).length,
    chatP50Ms: percentile(chat, 0.5),
    chatP95Ms: percentile(chat, 0.95),
    ttsP50Ms: percentile(tts, 0.5),
    ttsP95Ms: percentile(tts, 0.95),
    combinedP50Ms: percentile(combined, 0.5),
    combinedP95Ms: percentile(combined, 0.95),
    wallMs: Math.round((performance.now() - started) * 10) / 10,
  };
  levelSummaries.push(summary);
  console.log(JSON.stringify(summary));
}

const report = {
  gate: 5,
  topology: 'production Cloud Run /api/chat (Claude) -> exact PR2 Azure provider executed from GitHub Actions runner',
  productionTtsPrimaryChanged: false,
  firestoreSessionWriteCalls: 0,
  attemptedFlows: all.length,
  succeededFlows: all.filter((r) => r.ok).length,
  failedFlows: all.filter((r) => !r.ok).length,
  http429Total: all.filter((r) => r.chatStatus === 429 || r.error.includes('HTTP_429') || r.error.includes('HTTP 429')).length,
  providers: [...new Set(all.map((r) => r.provider).filter(Boolean))],
  voicesObserved: new Set(all.map((r) => r.voice).filter(Boolean)).size,
  levels: levelSummaries,
};

await fs.writeFile(`${outputDir}/gate5-summary.json`, JSON.stringify(report, null, 2));
await fs.writeFile(`${outputDir}/gate5-flows.json`, JSON.stringify(all, null, 2));

if (all.length !== levels.reduce((a, b) => a + b, 0)) throw new Error(`Gate 5 attempted flow count mismatch: ${all.length}`);
if (report.failedFlows !== 0) {
  const first = all.find((r) => !r.ok);
  throw new Error(`Gate 5 FAIL: ${report.failedFlows} flows failed; first=${JSON.stringify(first)}`);
}
if (report.http429Total !== 0) throw new Error(`Gate 5 FAIL: HTTP 429 count=${report.http429Total}`);
if (JSON.stringify(report.providers) !== JSON.stringify(['azure-speech'])) throw new Error(`Gate 5 FAIL: provider set=${JSON.stringify(report.providers)}`);
if (levelSummaries.some((row) => Number(row.combinedP95Ms || 0) > 30_000)) throw new Error('Gate 5 FAIL: combined p95 exceeded 30 seconds');

console.log('Gate 5: PASS');
console.log(JSON.stringify(report, null, 2));
